### PR TITLE
Support toggling static/dynamic plugin builds

### PR DESCRIPTION
**Pull request by dominiklohmann**
###  :notebook_with_decorative_cover: Description

This PR enables static plugin builds.

###  :memo: Checklist

- [x] All user-facing changes have changelog entries.
- [x] The changes are reflected on [docs.tenzir.com/vast](https://docs.tenzir.com/vast), if necessary.
- [x] The PR description contains instructions for the reviewer, if necessary.

### :dart: Review Instructions

Commit-by-commit. I suggest we peer-review this in a call because there are some things in here whose need may not be entirely obvious.

~This can already be reviewed before it's out of draft mode—the first commit contains the enabling CMake and plugin framework changes. Later commits simplify the plugin CMake.~ Ready for review.

### CODE REVIEW

File: vast/CMakeLists.txt [hunk 1]
```diff
@@ -11,6 +11,11 @@ add_executable(vast vast.cpp)
 target_link_libraries(vast PRIVATE vast::internal vast::libvast)
 install(TARGETS vast DESTINATION "${CMAKE_INSTALL_BINDIR}")
 
+# Use static versions of VAST_REGISTER_PLUGIN family of macros when inside the
+# vast binary. This makes it possible to statically link a plugin against the
+# vast::vast target, or to write built-in plugins that are always loaded with
+# VAST.
+target_compile_definitions(vast PRIVATE VAST_ENABLE_STATIC_PLUGINS_INTERNAL)
```
Comment (dominiklohmann) [inline vast/CMakeLists.txt:18]:
```suggestion
target_compile_definitions(vast PRIVATE VAST_ENABLE_STATIC_PLUGINS_INTERNAL)

```

File: vast/CMakeLists.txt [hunk 2]
```diff
@@ -141,7 +151,8 @@ if (VAST_PLUGINS)
   endforeach ()
 endif ()
 
-option(VAST_ENABLE_PLUGIN_AUTOLOADING "Always load all bundled plugins" OFF)
+cmake_dependent_option(VAST_ENABLE_PLUGIN_AUTOLOADING
+  "Always load all bundled plugins" OFF "NOT VAST_ENABLE_STATIC_PLUGINS" OFF)
```
Comment (dominiklohmann) [inline vast/CMakeLists.txt:155]: Make static plugins toggleable.

Comment (dominiklohmann) [reply vast/CMakeLists.txt:155]: Or write a follow-up story.

File: cmake/VASTRegisterPlugin.cmake [hunk 3]
```diff
@@ -0,0 +1,83 @@
+include_guard(GLOBAL)
+
+function (VASTRegisterPlugin)
+  include(GNUInstallDirs)
+
+  cmake_parse_arguments(PLUGIN "" "TARGET" "SOURCES;TEST_SOURCES" ${ARGN})
+
+  macro (target_link_whole_archive target visibility library)
```
Comment (dominiklohmann) [inline cmake/VASTRegisterPlugin.cmake:8]: Explain in comment not to ever touch this code.

File: libvast/src/plugin.cpp [hunk 4]
```diff
@@ -122,6 +119,15 @@ plugin_ptr::make(const char* filename, caf::actor_system_config& cfg) noexcept {
     // ranges overlap. Since this is static for the whole process, we just store
     // the already registed ID blocks from plugins in a static variable.
     static auto old_blocks = std::vector<::vast::plugin_type_id_block>{};
+    // Static plugins are built as part of the vast binary rather then libvast,
+    // so there will be runtime errors when there is a type ID clash between
+    // static and dynamic plugins. We register the ID blocks of all static
+    // plugins exactly once to always prefer them over dynamic plugins.
+    static auto flag = std::once_flag{};
+    std::call_once(flag, [&] {
+      for (const auto& [block, _] : plugins::get_type_id_blocks())
```
Comment (dominiklohmann) [inline libvast/src/plugin.cpp:128]:
```suggestion
      for (const auto& [block, _] : plugins::get_static_type_id_blocks())
```

File: libvast/src/plugin.cpp [hunk 5]
```diff
@@ -145,19 +155,22 @@ plugin_ptr::~plugin_ptr() noexcept {
     dlclose(library_);
     library_ = {};
   }
+  version_ = {};
```
Comment (dominiklohmann) [inline libvast/src/plugin.cpp:158]: Rework destructor logic for static plugins

**Review by dominiklohmann — COMMENTED** [review]
Peer-reviewed with @tobim.

File: vast/CMakeLists.txt (hunk 2 above)
Comment (dominiklohmann) [reply vast/CMakeLists.txt:156]: I created a follow-up story for discussion.

File: libvast/src/plugin.cpp (hunk 5 above)
Comment (dominiklohmann) [reply libvast/src/plugin.cpp:158]: The logic was right. I confused `instance_` and `library_` in the peer-review when looking at this.

**Review by tobim — APPROVED** [review]
Tested locally. Verified that type id clashes of static plugins result in a readable linker error.

**Comment by dominiklohmann** [comment]
I only rebased to fix the merge conflict in the CHANGELOG entry after approval.

**Pull request merged**